### ISSUE TITLE
Normalization for Optional/build doesn't always match semantics

### ISSUE THREAD

**Issue by ocharles** [dhall-lang/dhall-haskell#291]
I may be understanding something wrong, but I'm implementing

```
f ⇥ Optional/build A
g (Optional A) (λ(a : A) → [ a ] : Optional A) ([] : Optional A) ⇥ b
────────────────────────────────────────────────────────────────────
f g ⇥ b
```

The tests for this project show (in `normalization/optionalBuildA` and `normalization/optionalBuildB`) that

```
      λ(id : ∀(a : Type) → a → a)
    → Optional/build
      Bool
      (   λ(optional : Type)
        → λ(just : Bool → optional)
        → λ(nothing : optional)
        → id optional (just True)
      )
```

should reduce to

```
      λ(id : ∀(a : Type) → a → a)
    → Optional/build
      Bool
      (   λ(optional : Type)
        → λ(just : Bool → optional)
        → λ(nothing : optional)
        → id optional (just True)
      )
```

However, my implementation of the rule shows that it actually reduces to

```
λ(id : ∀(a : Type) → a → a) → id (Optional Bool) ([ True ] : Optional Bool)
```

I believe this is correct, as we can do this in `dhall`:

```
[nix-shell:~/work/dhall.js]$ dhall
\ ( id : forall ( a : Type ) -> a -> a ) ->
let A : Type = Bool in
let g = 
      (   λ(optional : Type)
        → λ(just : Bool → optional)
        → λ(nothing : optional)
        → id optional (just True)
      )
in g (Optional A) (λ(a : A) → [ a ] : Optional A) ([] : Optional A)

∀(id : ∀(a : Type) → a → a) → Optional Bool

λ(id : ∀(a : Type) → a → a) → id (Optional Bool) ([ True ] : Optional Bool)
```

Here I've bought `A` and `g` into scope with `let`. I know that `f ⇥ Optional/build A` holds, so I need to know what `b` is in `g (Optional A) (λ(a : A) → [ a ] : Optional A) ([] : Optional A) ⇥ b` - the above shows me it's just `id (Optional Bool) ([ True ] : Optional Bool)`.

`dhall.js` has this normalization behavior, but causes it to fail tests.

**Comment by ocharles** [comment]
Removing `id` does cause normalization:

```
Optional/build
      Bool
      (   λ(optional : Type)
        → λ(just : Bool → optional)
        → λ(nothing : optional)
        → just True
      )
Optional Bool

[ True ] : Optional Bool
```

I don't see why the presence of `id` should change things. Am I missing something?

**Comment by Gabriel439** [comment]
The spec is right, but the test and Haskell implementation are wrong, which I will fix

Regarding the latter question: that behavior you see is correct and expected.  `id` blocks normalization because the compiler doesn't know that it is the identity function.  It could in theory deduce that from the type of `id` since Dhall is total but codifying that sort of reasoning in general is tricky

**Comment by ocharles** [comment]
I didn't mean that the compiler should normalize `id` as identity, I meant I wasn't sure why `id` was causing the normalization behavior I was seeing above. But my reasoning was correct, and the Haskell implementation was wrong - that's what I was getting at.

**Comment by Gabriel439** [comment]
Actually, after digging into this more, I realize that both the semantics and the Haskell implementation need to be fixed.

The first thing I realized is that the standard semantics for `Optional/build` make `Optional/build`+`Optional/fold` fusion never trigger in one direction.  Specifically, in this example from the test suite:

```haskell
      λ ( f
        :   ∀(optional : Type)
          → ∀(just : Text → optional)
          → ∀(nothing : optional)
          → optional
        )
    → Optional/fold Text (Optional/build Text f)
```

... the standard semantics currently guarantees that `Optional/build Text f` will always reduce before fusion has a chance to trigger.  That's fine because it's not really that useful of a fusion rule (the other direction is the one that matters in practice).  Same issue with `List/fold`+`List/build` fusion.

The bigger issue with the standard semantics is that they are slightly wrong!  Specifically, they break in the corner case where the type variable is named `a`:

```haskell
λ(a : Type) → λ(g : ...) → Optional/build a g
```

... because that reduces to:

```haskell
λ(a : Type) → λ(g : ...) → g (Optional a) (λ(a : a) → [ a ] : Optional a) ([] : Optional a) ⇥ b
```

... which is a type error due to the fact that the `a` in `: Optional a` is not correctly shifted (i.e. it should be `a@1`).

So I will update the semantics to fix those issues and still update the Haskell implementation to match the semantics.

**Comment by Gabriel439** [comment]
Fixed by #300
